### PR TITLE
Feature/change download mechanism

### PR DESCRIPTION
**Pull request by redhug1**
### What

Utilise code from dp-download-service to facilitate encrypted download.
Add comments.
Refactor variable names to make code clearer to read for maintenance.
Get column width setting working.
Code clean up.

### How to review

Visual inspection.
Run tests.

### Who can review

Anyone.

### CODE REVIEW

File: handler/handler.go [hunk 1]
```diff
@@ -374,22 +420,27 @@ func (h *CsvComplete) GetCSVtoExcelStructure(ctx context.Context, excelInMemoryS
...
-			return &Error{err: err}
-		}
-		err = excelInMemoryStructure.SetColWidth(sheet1, "C", "C", 40) //!!! this is for test and needs further work to apply desired widths for all columns
-		if err != nil {
-			return &Error{err: err}
+		// Process and apply column widths
+		for i := 0; i < maxSettableColumnWidths; i++ {
+			if columnWidths[i] != columNotSet {
+				width := columnWidths[i] + 1 // add 1 to achieve slight visual space between columns and/or the vertical column lines
+				if width > 255 {
```
Comment (DavidSubiros) [inline handler/handler.go:431]: maybe we should create a constant (named `maxWidth` or something simliar) of value 255 and use it here?

File: handler/handler.go [hunk 2]
```diff
@@ -652,8 +703,10 @@ func generateURL(downloadServiceURL, instanceID string) string {
 
 // generateS3FilenameCSV generates the S3 key (filename including `subpaths` after the bucket)
 // for the provided instanceID CSV file that is going to be read
-func generateS3FilenameCSV(e *event.CantabularCsvCreated) string {
-	return fmt.Sprintf("instances/%s.csv", e.InstanceID)
+// !!! TODO filename should be datasets/<dataset_name>_<version>.csv to match CMD naming ???
+func generateS3FilenameCSV(event *event.CantabularCsvCreated) string {
```
Comment (DavidSubiros) [inline handler/handler.go:707]: this should match https://github.com/ONSdigital/dp-cantabular-csv-exporter/blob/develop/handler/handler.go#L358

```
datasets/%s-%s-%s.csv", e.DatasetID, e.Edition, e.Version
```

File: handler/handler.go [hunk 3]
```diff
@@ -663,15 +716,10 @@ func generateS3FilenameCSV(e *event.CantabularCsvCreated) string {
...
-func generateVaultPathForFile(vaultPathRoot string, e *event.CantabularCsvCreated) string {
-	return fmt.Sprintf("%s/%s.csv", vaultPathRoot, e.InstanceID)
-}
-
 // generateS3FilenameXLSX generates the S3 key (filename including `subpaths` after the bucket)
 // for the provided instanceID XLSX file that is going to be written
-func generateS3FilenameXLSX(e *event.CantabularCsvCreated) string {
-	return fmt.Sprintf("instances/%s.xlsx", e.InstanceID)
+func generateS3FilenameXLSX(event *event.CantabularCsvCreated) string {
+	return fmt.Sprintf("instances/%s.xlsx", event.InstanceID)
```
Comment (DavidSubiros) [inline handler/handler.go:722]: This should also have the name following the other files :
```
datasets/%s-%s-%s.csv", e.DatasetID, e.Edition, e.Version
```

File: handler/handler.go [hunk 4]
```diff
@@ -520,7 +571,7 @@ func (h *CsvComplete) UploadXLSXFile(ctx context.Context, e *event.CantabularCsv
 				)
 			}
 
-			vaultPath := generateVaultPathForFile(h.cfg.VaultPath, e)
+			vaultPath := fmt.Sprintf("%s/%s.xlsx", h.cfg.VaultPath, event.InstanceID)
```
Comment (DavidSubiros) [inline handler/handler.go:574]: if we change the filenames, this should match the new file name format (as expected by download service):
```
%s/%s-%s-%s.xlsx", h.cfg.vaultPath, e.DatasetID, e.Edition, e.Version

```

File: service/initialise.go [hunk 5]
```diff
@@ -104,19 +104,45 @@ var GetS3Uploaders = func(cfg *config.Config) (private, public S3Uploader, err e
...
 
-	// !!! is seems odd that the following function does not return an error, like the previous one
-	public = dps3.NewUploaderWithSession(cfg.UploadBucketName, private.Session())
-	return private, public, nil
+	publicUploader = dps3.NewUploaderWithSession(cfg.PublicBucketName, privateUploader.Session())
+	return privateUploader, publicUploader, nil
+}
+
+// GetS3Downloaders creates the private and public S3 Downloaders using the same AWS session, or a local storage client if a non-empty LocalObjectStore is provided
+var GetS3Downloaders = func(cfg *config.Config) (privateDownloader, publicDownloader *dps3.S3, err error) {
```
Comment (DavidSubiros) [inline service/initialise.go:122]: I think the downloaders could share the same AWS session with the uploaders, maybe this method could accept the session, and we can pass it from the uploaders? (or the other way around) - also, the uploaders have a nested client, I think it should be possible to call Get functions from there as well?

File: handler/handler.go (hunk 1 above)
Comment (redhug1) [reply handler/handler.go:431]: OK done

**Review by DavidSubiros — APPROVED** [review]
approved - as discussed, the remaining comments will be addressed when the code moves forward to us the latest schema.

**Pull request merged**